### PR TITLE
feat: support TypeScript 5.1

### LINKED ISSUE #6934: TypeScript 5.1 Support
# Note: This doesn't Block You From TypeScript 5.1

**You can still use `typescript@>=5.1` with `@typescript-eslint/eslint-plugin` & `@typescript-eslint/parser`.**

However, there may be cases wherein your project OOMs - which is a regression preventing us from officially landing support.

Please be patient while we work through this with the TS team.

---

https://devblogs.microsoft.com/typescript/announcing-typescript-5-1-beta/

This issue is just to track all of the new features and their implementation state in this project.
As with all releases, we will not necessarily to support all features until closer to the full release when everything the features are stabilised.

**_Please be patient._**

# 🔲 ES2020 and Node.js 14.17 as Minimum Runtime Requirements

As part of our upcoming [v6 release](https://typescript-eslint.io/blog/announcing-typescript-eslint-v6-beta) we've bumped our minimum node to v16.

This does not impact us as it's up to the user to use a supported node version for their selected TS version.

# 🔲 lib.d.ts Updates

We will need to regenerate our types within `scope-manager`.

-----

## Other changes with no impact to us

- Easier Implicit Returns for `undefined`-Returning Functions
- Unrelated Types for Getters and Setters
- Decoupled Type-Checking Between JSX Elements and JSX Tag Types
- Namespaced JSX Attributes
     - [TS already supported parsing namespaced attributes](https://typescript-eslint.io/play/#ts=5.0.3&jsx=true&sourceType=module&showAST=es&code=JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQduEAdqvAB5wC8cAPACbAAbnAIQIALgBGyKNzgAiAIwK4AegB8QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA) and we have AST types for it - so this is a no-op for us.
- `typeRoots` Are Consulted In Module Resolution
- Linked Cursors for JSX Tags
- Snippet Completions for `@param` JSDoc Tags
- Avoiding Unnecessary Type Instantiation
- Negative Case Checks for Union Literals
- Reduced Calls into Scanner for JSDoc Parsing
- Explicit `typeRoots` Disables Upward Walks for `node_modules/@types`

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
# Note: This doesn't Block You From Linting TypeScript 5.1

**You can still use `typescript@>=5.1` with `@typescript-eslint/eslint-plugin` & `@typescript-eslint/parser`.** There are no breaking changes in TS 5.1 that would block you from linting existing code. JSX namespaces (`<MyComponent a:b />`) should generally be linted properly.

Until this PR is merged and a new set of typescript-eslint package versions are released, your console may see a log like _`WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.`_.

## Status Update: June 26th 2023

Looks like TypeScript's latest releases have fixed the issue. We should be ready to release a new version with this PR's changes soon - pending review.

## Status Update: June 9 2023

This PR is blocked on out-of-memory (OOM) errors running a set of important unit tests for the parser. The OOM error only occurs with TypeScript 5.1, not locally.

We're working on identifying which commit to TypeScript introduced the change that coincides with this OOM error being introduced. In https://github.com/typescript-eslint/typescript-eslint/pull/7091 we narrowed it down as follows:

* ✅ Running with `typescript@5.1.0-dev.20230301` _does not_ exhibit the OOM error
* ❌ Running with `typescript@5.1.0-dev.20230302` **does** exhibit the OOM error

~Our next step will be to write a script to run against individual TypeScript commits.~

It looks like https://github.com/microsoft/TypeScript/issues/54542 is the root of the OOM. At least, from running on recent TypeScript commits, that's the first one that has the OOM occur. TBD once it's fixed.

> Thanks to @jakebailey from the TypeScript team for helping us with this! ❤️ 

---

## PR Checklist

- [x] Addresses an existing open issue: fixes #6934
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Bumps the supported ranges per https://typescript-eslint.io/maintenance/versioning/dependant-version-upgrades#adding-support-for-a-new-typescript-version. Since we didn't have an RC PR, this merges the PR steps a bit.

### CODE REVIEW

File: packages/typescript-estree/src/convert.ts
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/src/convert.ts:585]: > [TS already supported parsing namespaced attributes](https://typescript-eslint.io/play/#ts=5.0.3&jsx=true&sourceType=module&showAST=es&code=JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQduEAdqvAB5wC8cAPACbAAbnAIQIALgBGyKNzgAiAIwK4AegB8QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA) and we have AST types for it - so this is a no-op for us.

I think this is a bit more efficient of a transform. But I'm a little lost in the using `JSXIdentifier` vs `Identifier` - do we want to normalize one way or another?

File: patches/typescript+5.1.3.patch
Comment (JoshuaKGoldberg) [inline patches/typescript+5.1.3.patch:1]: I'm between airports and don't have great wifi at the moment - if someone could recreate this for me, that'd be great...

Comment (JoshuaKGoldberg) [reply patches/typescript+5.1.3.patch:1]: Thank you, free Sydney airport wifi!

File: packages/ast-spec/tsconfig.build.json
Comment (JoshuaKGoldberg) [inline packages/ast-spec/tsconfig.build.json:5]: Otherwise:

```plaintext
> nx run ast-spec:typecheck

$ tsc -p tsconfig.json --noEmit
src/jsx/JSXNamespacedName/fixtures/component/fixture.tsx:1:11 - error TS17004: Cannot use JSX unless the '--jsx' flag is provided.

1 const x = <NamespacePropComponent a:b="tight spacing" />;
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

https://github.com/typescript-eslint/typescript-eslint/actions/runs/5194185770/jobs/9365660553?pr=7088

**Comment by JoshuaKGoldberg** [comment]
```plaintext
Run ./.github/actions/wait-for-netlify
/home/runner/work/typescript-eslint/typescript-eslint/node_modules/@actions/core/lib/core.js:106
        throw new Error(`Input required and not supplied: ${name}`);
        ^

Error: Input required and not supplied: netlify_token
```

@JamesHenry something funky with the new action, perhaps? https://github.com/typescript-eslint/typescript-eslint/actions/runs/5195237524/jobs/9367800550?pr=7088

**Comment by JoshuaKGoldberg** [comment]
```plaintext
PASS  tests/lib/node-utils.test.ts
 PASS  tests/lib/parse.project-true.test.ts
<--- Last few GCs --->
[2972:0x61b15c0]   201164 ms: Mark-sweep (reduce) 2037.5 (2051.0) -> 2036.7 (2052.5) MB, 2662.2 / 0.0 ms  (average mu = 0.171, current mu = 0.014) allocation failure scavenge might not succeed
[2972:0x61b15c0]   203868 ms: Mark-sweep (reduce) 2037.8 (2051.5) -> 2037.3 (2052.8) MB, [26](https://github.com/typescript-eslint/typescript-eslint/actions/runs/5195237524/jobs/9367799309?pr=7088#step:6:27)32.6 / 0.0 ms  (average mu = 0.102, current mu = 0.026) allocation failure scavenge might not succeed
<--- JS stacktrace --->
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xa3ad50 node::Abort() [/opt/hostedtoolcache/node/14.21.3/x64/bin/node]
```

_sigh_.

File: packages/typescript-estree/src/convert.ts
Comment (bradzacher) [reply packages/typescript-estree/src/convert.ts:585]: You updated your comment but to answer your previous question...

> But I'm a little lost in the using JSXIdentifier vs Identifier - do we want to normalize one way or another?

JSX identifier exists because [JSX is specced to allow names with `-`](https://facebook.github.io/jsx/#prod-JSXIdentifier).

EG `<foo-bar />` creates a `JSXIdentifier { name = 'foo-bar' }` - and this name is obviously is not a valid `Identifier`.
So we cannot normalise one way or the other, sadly 😢

Comment (bradzacher) [reply packages/typescript-estree/src/convert.ts:585]: Yeah it's a good idea to be more explicit if we can considering that now TS has an explicit node for namespacing.

File: packages/scope-manager/tests/fixtures/jsx/namespaced-attribute.tsx
Comment (bradzacher) [inline packages/scope-manager/tests/fixtures/jsx/namespaced-attribute.tsx:5]: cool - this case actually used to be an error pre 5.1!

**Review by bradzacher — APPROVED** [review]
One case that would be worth adding a test for:
`<foo-bar:baz-bam />`

Both the ns and name of the namespace are allowed to be valid JSX identifiers!

Otherwise this is all LGTM
Once you land it we can do an out-of-band release to get it out and keep the early adopters happy!

File: packages/typescript-estree/src/convert.ts
Comment (JoshuaKGoldberg) [reply packages/typescript-estree/src/convert.ts:585]: Ha yeah I figured part of it out a bit after commenting. This makes sense, thanks!

**Comment by JamesHenry** [comment]
> ```
> Run ./.github/actions/wait-for-netlify
> /home/runner/work/typescript-eslint/typescript-eslint/node_modules/@actions/core/lib/core.js:106
>         throw new Error(`Input required and not supplied: ${name}`);
>         ^
> 
> Error: Input required and not supplied: netlify_token
> ```
> 
> 
>     
>   
> 
> @JamesHenry something funky with the new action, perhaps? https://github.com/typescript-eslint/typescript-eslint/actions/runs/5195237524/jobs/9367800550?pr=7088

Oh sorry, I reckon it could be that the secret doesn’t work as an input when you’re coming from a fork…

**Comment by JamesHenry** [comment]
I thought that should work based on this answer: https://github.com/orgs/community/discussions/26409#discussioncomment-3251822

anyone know the right syntax here?

**Comment by bradzacher** [comment]
@JoshuaKGoldberg once you've got all the tests passing locally feel free to just land this to main.

We can follow up with a fix commit if any CIs fail on main.

File: packages/typescript-estree/jest.config.js
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/jest.config.js:16]: 😬 I had hoped that it was just Node 14 running out of memory, but no - 18 runs out too. https://github.com/typescript-eslint/typescript-eslint/actions/runs/5206328809/jobs/9392825081

```plaintext
Running with v18.16.0 which startsWith: false
```

```
Start of describe: parse.moduleResolver.placeholder-error.test.ts

      at log (tests/lib/parse.moduleResolver.placeholder-error.test.ts:13:11)
<--- Last few GCs --->
[2965:0x6b05f90]   159828 ms: Mark-sweep (reduce) 2040.3 (2082.4) -> 2039.4 (2082.9) MB, 1581.5 / 0.0 ms  (average mu = 0.[261](https://github.com/typescript-eslint/typescript-eslint/actions/runs/5206328809/jobs/9392825081#step:5:262), current mu = 0.180) allocation failure; scavenge might not succeed
```

Comment (JoshuaKGoldberg) [reply packages/typescript-estree/jest.config.js:16]: I think we're being hit by https://github.com/microsoft/TypeScript/issues/54491.

**Comment by JoshuaKGoldberg** [comment]
Per https://github.com/typescript-eslint/typescript-eslint/pull/7088/files#r1222369425, marking this blocked on https://github.com/microsoft/TypeScript/pull/54536. Once `typescript@5.1.4` is released I can update this PR to use it. Hopefully that'll fix the issues.

**Comment by bradzacher** [comment]
@JoshuaKGoldberg did we want to test the on-demand for that PR to double check that it fixes it?

https://github.com/microsoft/TypeScript/pull/54536#issuecomment-1577503845
```
{
    "devDependencies": {
        "typescript": "https://typescript.visualstudio.com/cf7ac146-d525-443c-b23c-0d58337efebc/_apis/build/builds/155350/artifacts?artifactName=tgz&fileId=25390837027CDB2F59440AFD227BEAC42D7474DF7E5DDBB45B82274982C27B3002&fileName=/typescript-5.2.0-insiders.20230605.tgz"
    }
}
```

I'd assume it's the same bug but probs worth verifying?

File: .github/actions/prepare-install/action.yml
Comment (JoshuaKGoldberg) [inline .github/actions/prepare-install/action.yml:58]: https://github.com/typescript-eslint/typescript-eslint/actions/runs/5207310238/jobs/9394739015

```plaintext
warning Lockfile has incorrect entry for "typescript@>=3.3.1 <5.2.0". Ignoring it.
error Your lockfile needs to be updated, but yarn was run with `--frozen-lockfile`.
```

I don't get this locally 😡

**Comment by bradzacher** [comment]
looks like it's still OOMing even with that patch version 😢

**Comment by JoshuaKGoldberg** [comment]
Yeah 😞 good callout on trying things here. I still want to wait until the OOM fix lands, since it looks like it'll be a quick follow and I don't have much bandwidth for a deep investigation right now.

**Comment by JoshuaKGoldberg** [comment]
Ah, this is from on a _different_ TypeScript issue likely to be fixed soon: https://github.com/microsoft/TypeScript/issues/54542

File: packages/typescript-estree/tests/fixtures/simpleProject/tsconfig.json
Comment (JoshuaKGoldberg) [reply packages/typescript-estree/tests/fixtures/simpleProject/tsconfig.json:6]: Correction: https://github.com/microsoft/TypeScript/issues/54542

File: package.json
Comment (JoshuaKGoldberg) [reply package.json:114]: @jakebailey indicated it should be patched soon

Comment (jakebailey) [reply package.json:114]: Yeah there will be a patch release of 5.1 in the next day or so.

File: packages/typescript-estree/jest.config.js
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/jest.config.js:5]: oops, meant to remove this... will do later 😄

**Pull request merged**
